### ISSUE TITLE
Squash option when merging a branch

### ISSUE THREAD

**Issue by benrubson** [desktop/desktop#2507]
Hi,

When we merge a branch (`Branch / Merge Into Current Branch...`), would be nice to have an option to squash (`git merge --squash`).

A simple option **_squash_** to tick would be perfect, internally it would then simply add `--squash` to the git command.

Many thanks for your support 👍 

Ben

**Comment by shiftkey** [comment]
@benrubson thanks for the suggestion!

This idea is interesting for the future, but this is beyond the scope of our [current roadmap](https://github.com/desktop/desktop/blob/master/docs/process/roadmap.md).

I added the `future-proposal` label to this issue and closed it so that we won't get overwhelmed with our backlog! We will revisit this in our next planning session.

**Comment by benrubson** [comment]
Many thx @shiftkey 👍

**Comment by davidcie** [comment]
Would love to see this option come to life too! I know this is a matter of taste but some of us commit a lot but prefer not to clutter master too much with the entire history of small changes.

**Comment by steveward** [comment]
@MCrafterzz requested this in https://github.com/desktop/desktop/issues/8569.

>GitHub desktop makes git much easier to use and makes stuff like pushing, pulling and branch creation trivial. But on feature I really miss from git is squashing commits together (git rebase -I HEAD~2) which is essential in some cases, for example contributing to the godot game engine.

**Comment by 29er** [comment]
It's such a bummer we cant do this in desktop. command line it is !

**Comment by hyadav5** [comment]
Is this feature still waiting for the day of light.

**Comment by JosephTLyons** [comment]
> It's such a bummer we cant do this in desktop. command line it is !

GitHub.com has it too....

**Comment by RedDevilus** [comment]
![image](https://user-images.githubusercontent.com/24227051/99105294-8cccca00-25e2-11eb-8435-fafde5f657c1.png)

It's basically a similar way on how to do interactive with git and a text editor but more on the GUI level.
(Yes i know drop commit isn't on there but it's just a sketch.)

**Comment by usmcmckee** [comment]
Now March 12th 2021 almost 4 years later. Is the squash feature request going to be added to the roadmap?

**Comment by billygriffin** [comment]
Hi folks, we're starting to look at this so I'm going to reopen this issue. For merging, I think we'll likely do something similar to what we did with rebasing, where it's a separate path instead of taking over the existing merge path. We're also hoping to implement something to allow you to squash commits on a particular branch outside the context of a merge. If you've 👍 on this issue and wouldn't mind sharing your use cases for when you're squashing, that would be really helpful! We may also reach out here and ask for feedback and/or usability testing participants once we've shipped an initial iteration to beta. Thanks!

**Comment by AFDevMike** [comment]
Our use case is: 

- Developer works on some task by checking out a task branch from `dev`.
- They are free to save commits on that task branch willy nilly with names like `save`, or `now builds`.
- When they are finished, they check out `dev` branch, pull  and `git merge --squash` the temp task branch. This produces 1 clean commit per task.

**Comment by JosephTLyons** [comment]
> Our use case is:
> 
>     * Developer works on some task by checking out a task branch from `dev`.
> 
>     * They are free to save commits on that task branch willy nilly with names like `save`, or `now builds`.
> 
>     * When they are finished, they check out `dev` branch, pull  and `git merge --squash` the temp task branch. This produces 1 clean commit per task.

I do this procedure with my personal projects.

**Comment by billygriffin** [comment]
Thanks @AFDevMike, I think the biggest gap in my understanding is why that squash + merge to dev is happening locally as opposed to in a pull request on GitHub.com. Would you mind sharing context around merging locally vs remotely and how that works for y'all?

**Comment by usmcmckee** [comment]
@billygriffin, I know from my experience that not every company has a good PR policy. I am a Senior Consultant, and I have seen a different strategy for merging with each of my clients. My current client has two dev branches and only PR into the higher branch but merges into the lower branch. We use Github Desktop with their Azure DevOps environment. They will be switch to GitHub Enterprise next year. It would be nicer to use GitHub Desktop instead of Git Bash to help keep a cleaner history and fewer commits on the lower dev branch.

**Comment by AFDevMike** [comment]
In our workflow, the developer pushes (to remote dev) after merge.  I don't have a well reasoned answer to why we don't use pull requests for our development.  Maybe we need better PR hygiene? :)  We have a process for Testing before pushing and its worked for us but we are a 2 person team.  I'll love to re-evaluate our process and even set up CI/CD but we are in Crunch until next main product releases.  Recently added a new Dev who wanted to use GitHub desktop and this was the 1 feature they couldn't accomplish wo dropping down to Git Bash.

**Comment by RedDevilus** [comment]
Git bash mingw to then do git rebase -i HEAD~5 as example to fix mistakes like having a good first commit, second commit is bad and the rest are good. So if there is an easy way to do revert till second commit or do a fix commit and squash to 5 commits again. Esspecialy handy if you use a clang format where you can easily make things difficult for yourself. About the only feature to me that github desktop lacks in. Interactive rebasing or even just squash merge.

**Comment by tidy-dev** [comment]
Adding a work scenario I have had that could have used this feature. I have been on a team that required you to commit and push your current work up at the end of every day (actually was preferred midday and end of day). This resulted in several wip commits which I don't particularly care for (as could mean commit with code in a broken state). Thus, to work with that I would create a feature-A branch and branch off of that with wip commits and then merge my wip branch into my feature-A after that sub section was finished/functional/non-breaking. Tho I didn't squash merge them, this feature could be useful here.  

That said.. 😄  I don't know that this is a common practice.

**Comment by billygriffin** [comment]
Thanks y'all, this is all really helpful. ❤️ If others want to weigh in with more context on your workflows, especially if they differ from what others described above, please do!

**Comment by usmcmckee** [comment]
@billygriffin I would recommend checking out GitLen's extension in VS Code. You can rebase squash easily with this tool even on the same branch. Here are some ideas. 
![image](https://user-images.githubusercontent.com/52210324/117495180-e7fccb00-af3a-11eb-956d-000fdfe0dfb5.png)
![image](https://user-images.githubusercontent.com/52210324/117495194-eb905200-af3a-11eb-8b79-bff7b85199b9.png)

**Comment by Kurt-von-Laven** [comment]
This is the git/GitHub workflow I was taught by Linux kernel developers, the original purpose for which git was designed. It has a bit of a learning curve, but produces a commit history of the highest quality when applied correctly. With a bit of practice my experience has been that it actually accelerates even my initial authoring of correct software as opposed to slowing me down or distracting from feature development. The practice of decomposing my changes into correct commits and authoring helpful commit messages often reveals my own mistakes to me. I don't recommend this workflow, however, if your goal is to hastily churn out buggy code; nor do I recommend attempting to transition to it during a period of unusual organizational stress. When the commits are clean and coherent, it facilitates code review, maintenance (especially when an engineer leaves the organization), debugging with `git bisect`, the use of `git blame` to offer context, etc. I would love to see interactive rebases supported by GitHub Desktop so that engineers who are less familiar with git and the command line can devote less of their focus to ramping up on the git CLI. In the meantime, I will encourage use of [GitLens in VSCode](https://marketplace.visualstudio.com/items?itemName=eamodio.gitlens#interactive-rebase-editor), but that has the drawback of requiring people to either use two different editors or leave their editor of choice. I kept the explanations of the commands as a reminder of how much newcomers to git are already being asked to learn at once.

1. `git checkout main` (Switch to the main branch if you aren't already on it.)
2. `git pull` (Update the main branch with any changes to the main branch on GitHub.)
3. `git checkout -b <branch_name>` (Make a new branch and switch to it.)
4. Make the pertinent changes to your task. (It's fine to fix other things along the way within reason if desired, but put them in separate commits.)
5. `git add -p <modified_file>` (Select the appropriate hunks to stage from each file you made changes to, repeating as needed. This process is invaluable for self review as we are psychologically more likely to detect our own mistakes when presented with them in a different format.)
6. `git commit`  (Repeat steps 5-6 once per logical coherent describable unit of work; all of the changes within one commit should relate to one another.)
7. `git push --set-upstream origin <branch_name>` (Copy your branch to GitHub. Aim for fewer than 100 to 300 lines per pull request, which is all the commits you push in one unit for review.)
8. Open a pull request to merge your branch to main at GitHub. (The output of git push contains a link to do this.)
9. Address review feedback, if any.
10. `git add -p <modified_file>` (This is the same as step 5, but the changes will be relative to the changes you previously made.)
11. `git commit` (Add the `--amend` flag if there is only one commit in your pull request to squash all your changes together and avoid polluting the revision history with a bunch of commits that say "addressed code review feedback.")
12. `git rebase -i main` (If there is more than one commit in your pull request, you can use this command to fixup specific commits in response to your review feedback.)
13. `git push --force-with-lease` (Overwrite your branch on GitHub with the latest and greatest version of your branch.)
14. Once your pull request is approved, rebase and merge your branch into main using the green button at the bottom, and delete the branch on GitHub.
15. `git checkout main` (Switch to the main branch.)
16. `git pull` (Update the main branch with the contents of your recently merged pull request and any other changes that may have landed in the meantime.)
17. `git checkout <branch_name>` (Switch to the feature branch.)
18. `git rebase main` (Rebase the feature branch to ensure that you merged what you expected to; this should succeed without intervention.)
19. `git checkout main` (Switch back to main.)
20. `git branch -d <branch_name>` (Delete the local branch now that it's merged and rebased.)

**Comment by goyalyashpal** [comment]
It's in [2.8.4-beta1](https://desktop.github.com/release-notes/?env=beta) June 3rd 2021: https://github.com/desktop/desktop/pull/12336

**Update:** In [2.9.0 Stable](https://desktop.github.com/release-notes/) June 16th 2021

**Comment by totoprayogo1916** [comment]
after squash: 
![image](https://user-images.githubusercontent.com/2387514/122503046-edf9c780-d021-11eb-9da5-af7d88793ab6.png)

The `date` is not updated,

**Comment by tidy-dev** [comment]
@totoprayogo1916 This is expected because what we show for that date is the author date. During a squash the interactive rebase retains the oldest author date of the squashed commits. If you were to go to your terminal and run `git log --pretty=fuller` and compare the `CommitDate` to the `AuthorDate`, you would see the `CommitDate` would be the day you squashed it and the `AuthorDate` to be the earliest some of the code was first committed.

**Comment by Nixinova** [comment]
> `CommitDate` would be the day you squashed it and the `AuthorDate` to be the earliest some of the code was first committed.

This can be a bit annoying because GitHub only reads the commit date, so if you squash two commits from a while ago all subsequent commits are then redated to today according to GitHub history.

**Comment by bungfut** [comment]
RFC2617
![9r](https://user-images.githubusercontent.com/33502002/124957589-711ba580-e043-11eb-98f6-d3db4f988306.png)

**Comment by bungfut** [comment]
k.futwatchara@gmail.com

**Comment by codedfiles** [comment]
explainers.md

**Comment by Case244** [comment]
> Hi,
> 
> 
> 
> When we merge a branch (`Branch / Merge Into Current Branch...`), would be nice to have an option to squash (`git merge --squash`).
> 
> 
> 
> A simple option **_squash_** to tick would be perfect, internally it would then simply add `--squash` to the git command.
> 
> 
> 
> Many thanks for your support 👍 
> 
> 
> 
> Ben
> 
> 

- - [ ]

**Comment by xayotdankonglimqoldi** [comment]
telegram is very good
